daisi/src/cpps/logical/amr/amr_logical_agent.cpp

```diff
@@ -0,0 +1,219 @@
+                   first_node),
+      description_set_(false),
+      topology_(daisi::util::Dimensions(50, 20, 0)),
+      current_state_(AmrState::kIdle)  // TODO placeholder
+{}
+
+void AmrLogicalAgent::init(ns3::Ptr<ns3::Socket> tcp_socket) {
+  initCommunication();
+
+  server_socket_ = tcp_socket;
+  server_socket_->Listen();
+
+  server_socket_->SetAcceptCallback(
+      ns3::MakeCallback(&AmrLogicalAgent::connectionRequest, this),
+      ns3::MakeCallback(&AmrLogicalAgent::newConnectionCreated, this));
+}
+
+void AmrLogicalAgent::initAlgorithms() {
+  if (!description_set_) {
+    throw std::runtime_error("Initialization not finished yet.");
+  }
+
+  // TODO decide about chosen order management depending on algorithm config
+  order_management_ = std::make_shared<order_management::StnOrderManagement>(
+      description_, topology_, daisi::util::Pose{current_position_});
+
+  for (const auto &algo_type : algorithm_config_.algorithm_types) {
+    switch (algo_type) {
+      case AlgorithmType::kDispositionParticipant:
+        algorithms_.push_back(std::make_unique<DispositionParticipant>(sola_));
+        break;
+      default:
+        throw std::invalid_argument("Algorithm Type cannot be initiated on Amr Logical Agent.");
+    }
+  }
+}
+
+void AmrLogicalAgent::start() { initAlgorithms(); }
+
+void AmrLogicalAgent::messageReceiveFunction(const sola::Message &msg) {
+  // TODO add logging of message
+  this->LogicalAgent::messageReceiveFunction(msg);
+}
+
+void AmrLogicalAgent::topicMessageReceiveFunction(const sola::TopicMessage &msg) {
+  // TODO add logging of topic message
+  this->LogicalAgent::topicMessageReceiveFunction(msg);
+}
+
+void AmrLogicalAgent::readFromPhysicalSocket(ns3::Ptr<ns3::Socket> socket) {
+  // TODO add logging?
+  // TODO outsource into algorithm; would that make sense?
+
+  auto packet = socket->Recv();
+  daisi::cpps::CppsTCPMessage header;
+  packet->RemoveHeader(header);
+
+  for (const auto &msg : header.getMessages()) {
+    auto amr_msg = amr::deserialize(msg.payload);
+
+    if (auto amr_description = std::get_if<AmrDescription>(&amr_msg)) {
+      processMessageAmrDescription(*amr_description);
+
+    } else if (auto amr_status_update = std::get_if<AmrStatusUpdate>(&amr_msg)) {
+      processMessageAmrStatusUpdate(*amr_status_update);
+
+    } else if (auto amr_order_update = std::get_if<AmrOrderUpdate>(&amr_msg)) {
+      processMessageAmrOrderUpdate(*amr_order_update);
+    }
+  }
+}
+
+void AmrLogicalAgent::processMessageAmrDescription(const AmrDescription &description) {
+  if (!description_set_) {
+    description_ = description;
+  } else {
+    throw std::runtime_error("AmrDescription attempted to be set twice.");
+  }
+}
+
+void AmrLogicalAgent::processMessageAmrStatusUpdate(const AmrStatusUpdate &status_update) {
+  current_position_ = status_update.getPosition();
+  current_state_ = status_update.getState();
+
+  checkSendingNextTask();
```

Comment:
Will you send the next task if the previous task is not finished yet? `AmrStatusUpdate` does not send information about the task.